src/Compiler/lex.fsl

```diff
@@ -1030,63 +1031,45 @@ rule token (args: LexArgs) (skip: bool) = parse
+     let tok = HASH_IF(m, lexed, LexCont.EndLine(args.ifdefStack, args.stringNest, contCase))
+     if skip then endline contCase args skip lexbuf else tok }
 
-     // Get the token; make sure it starts at zero position & return
-     let cont, f =
-         if isTrue then
-             let cont = LexCont.EndLine(args.ifdefStack, args.stringNest, LexerEndlineContinuation.Token)
-             let f = endline LexerEndlineContinuation.Token args skip
-             cont, f
-         else
-             let cont = LexCont.EndLine(args.ifdefStack, args.stringNest, LexerEndlineContinuation.Skip(0, m))
-             let f = endline (LexerEndlineContinuation.Skip(0, m)) args skip
-             cont, f
-
-     let tok = shouldStartLine args lexbuf m (FSComp.SR.lexHashIfMustBeFirst()) (HASH_IF(m,lexed,cont))
-     if not skip then tok else f lexbuf }
-
- | anywhite* "#else" anywhite* ("//" [^'\n''\r']*)?
+ | anywhite* "#else" anywhite* ("//" anystring)?
    { let lexed = (lexeme lexbuf)
      match args.ifdefStack with
      | [] ->  LEX_FAILURE (FSComp.SR.lexHashElseNoMatchingIf())
      | (IfDefElse,_) :: _rest -> LEX_FAILURE (FSComp.SR.lexHashEndifRequiredForElse())
      | (IfDefIf,_) :: rest ->
        let m = lexbuf.LexemeRange
+       shouldStartLine args lexbuf m (FSComp.SR.lexHashElseMustBeFirst())
        args.ifdefStack <- (IfDefElse,m) :: rest
        LexbufIfdefStore.SaveElseHash(lexbuf, lexed, m)
-       let tok = HASH_ELSE(m, lexed, LexCont.EndLine(args.ifdefStack, args.stringNest, LexerEndlineContinuation.Skip(0, m)))
-       let tok = shouldStartLine args lexbuf m (FSComp.SR.lexHashElseMustBeFirst()) tok
-       if not skip then tok else endline (LexerEndlineContinuation.Skip(0, m)) args skip lexbuf }
+       let tok = HASH_ELSE(m, lexed, LexCont.EndLine(args.ifdefStack, args.stringNest, LexerEndlineContinuation.IfdefSkip(0, m)))
+       if skip then endline (LexerEndlineContinuation.IfdefSkip(0, m)) args skip lexbuf else tok }
 
- | anywhite* "#endif" anywhite* ("//" [^'\n''\r']*)?
+ | anywhite* "#endif" anywhite* ("//" anystring)?
    { let lexed = (lexeme lexbuf)
      let m = lexbuf.LexemeRange
+     shouldStartLine args lexbuf m (FSComp.SR.lexHashEndifMustBeFirst())
      match args.ifdefStack with
      | []->  LEX_FAILURE (FSComp.SR.lexHashEndingNoMatchingIf())
      | _ :: rest ->
         args.ifdefStack <- rest
         LexbufIfdefStore.SaveEndIfHash(lexbuf, lexed, m)
         let tok = HASH_ENDIF(m,lexed,LexCont.EndLine(args.ifdefStack, args.stringNest, LexerEndlineContinuation.Token))
-        let tok = shouldStartLine args lexbuf m (FSComp.SR.lexHashEndifMustBeFirst()) tok
         if not skip then tok else endline LexerEndlineContinuation.Token args skip lexbuf }
 
  | "#if"
    { let tok = WHITESPACE (LexCont.Token (args.ifdefStack, args.stringNest))
      let tok = fail args lexbuf (FSComp.SR.lexHashIfMustHaveIdent()) tok
-     if not skip then tok else token args skip lexbuf }
-
- | anywhite* "#if" ident_char+
```

Comment:
Wasn't this "lexer recovery" for common mistakes, e.g. not crashing the rest of the lexing/parsing when a whitespace is missing between a hash-keyword and an identifier?

Reply:
Oops, that sounds correct. I will revert it. Sorry for being to fast here.

Reply:
I wonder, though, why we don't just report an error here during lexing, rather than dealing with it four different files. But since this PR was meant to not change lexing, I will revert it.

Reply:
I think it keeps the token since every rule has to produce "something", so that other phases can continue.
Not sure if all other places handling it are necessary, but the lexer recovery and follow up code (parsing+typechecking in order to get colouring) does make sense to me.

Reply:
Yeah, the rule makes sure that `#ifxyz` and  `#xyz` are colored in the same way.